presto-main/src/main/java/com/facebook/presto/type/TimestampOperators.java

```diff
@@ -137,8 +137,17 @@ public static long castToTime(ConnectorSession session, @SqlType(StandardTypes.T
     @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
     public static long castToTimeWithTimeZone(ConnectorSession session, @SqlType(StandardTypes.TIMESTAMP) long value)
     {
-        int timeMillis = modulo24Hour(getChronology(session.getTimeZoneKey()), value);
-        return packDateTimeWithZone(timeMillis, session.getTimeZoneKey());
+        if (session.isLegacyTimestamp()) {
+            int timeMillis = modulo24Hour(getChronology(session.getTimeZoneKey()), value);
+            return packDateTimeWithZone(timeMillis, session.getTimeZoneKey());
+        }
+        else {
+            ISOChronology localChronology = getChronology(session.getTimeZoneKey());
+
+            // This cast does treat TIMESTAMP as wall time in session TZ. This means that in order to get
+            // its UTC representation we need to shift the value by the offset of TZ.
+            return packDateTimeWithZone(modulo24Hour(value - localChronology.getZone().getOffset(value)), session.getTimeZoneKey());
```

Comment:
Is this equivalent to casting to TIMESTAMP_WITH_TZ first and that casting to TIME_WITH_TZ by truncating y,d,m info? Do we have tests covering it is always so? Perhaps even regardless of legacy switch?